Index: kaffeine/src/kaffeine.desktop
===================================================================
--- kaffeine/src/kaffeine.desktop	(revision 708328)
+++ kaffeine/src/kaffeine.desktop	(working copy)
@@ -57,7 +57,7 @@
 GenericName[sv]=Mediaspelare
 GenericName[tg]=Навозишгари Медиа
 GenericName[th]=โปรแกรมเล่นสื่อ
-GenericName[tr]=Oynatıcı
+GenericName[tr]=Çokluortam Oynatıcısı
 GenericName[uk]=Програвач медіа-матеріалів
 GenericName[xx]=xxMedia Playerxx
 GenericName[zh_CN]=媒体播放器
