Index: src/player-parts/xine-part/xine_part.h
===================================================================
--- src/player-parts/xine-part/xine_part.h	(revision 746328)
+++ src/player-parts/xine-part/xine_part.h	(working copy)
@@ -108,6 +108,8 @@
 	QString screenShot();
 	void nextAudioChannel();
 	void nextSubtitleChannel();
+	void speedFaster();
+	void speedSlower();
 
 public slots:
 	/*
Index: src/player-parts/xine-part/xine_part_iface.h
===================================================================
--- src/player-parts/xine-part/xine_part_iface.h	(revision 746328)
+++ src/player-parts/xine-part/xine_part_iface.h	(working copy)
@@ -128,6 +128,8 @@
 
   virtual void nextAudioChannel() = 0;
   virtual void nextSubtitleChannel() = 0;
+  virtual void speedFaster() = 0;
+  virtual void speedSlower() = 0;
 };
 
 #endif /* XINE_PART_IFACE_H */
Index: src/player-parts/xine-part/xine_part.cpp
===================================================================
--- src/player-parts/xine-part/xine_part.cpp	(revision 746328)
+++ src/player-parts/xine-part/xine_part.cpp	(working copy)
@@ -449,6 +449,11 @@
 	}
 }
 
+void XinePart::speedFaster()
+{
+	slotFastForward();
+}
+
 void XinePart::slotFastForward()
 {
 	if (m_xine->getSpeed() == KXineWidget::Pause)
@@ -459,6 +464,11 @@
 	m_xine->slotSpeedFaster();
 }
 
+void XinePart::speedSlower()
+{
+	slotSlowMotion();
+}
+
 void XinePart::slotSlowMotion()
 {
 	if (m_xine->getSpeed() == KXineWidget::Pause)
