--- scandialog.cpp      2007-12-21 21:21:22.000000000 +0000
+++ newscandialog.cpp   2007-12-21 21:23:26.000000000 +0000
@@ -109,13 +109,15 @@
        foundList->clear();
        foundList->setAllColumnsShowFocus( true );
        foundList->setSelectionMode( QListView::Extended );
-       channelsList->setSorting( 0 );
+       channelsList->setSorting( 1 ); // sort by source 1st than channel name
+       channelsList->setAllColumnsShowFocus( true );
+       channelsList->setSelectionMode( QListView::Extended );

        ChannelDesc *chan;
        QListViewItem *it;
        for ( int i=0; i<(int)chandesc->count(); i++ ) {
                chan = chandesc->at(i);
-               it = new QListViewItem( channelsList, chan->name );
+               it = new QListViewItem( channelsList, chan->name, chan->tp.source );
                if ( !chan->pix.isNull() )
                        it->setPixmap( 0, chan->pix );
                else {

--- scandialogui.ui     2007-12-21 21:21:22.000000000 +0000
+++ newscandialogui.ui  2007-12-12 18:59:16.000000000 +0000
@@ -66,9 +66,28 @@
                                             <bool>true</bool>
                                         </property>
                                     </column>
+                                    <column>
+                                        <property name="text">
+                                            <string>Source</string>
+                                        </property>
+                                        <property name="clickable">
+                                            <bool>true</bool>
+                                        </property>
+                                        <property name="resizable">
+                                            <bool>true</bool>
+                                        </property>
+                                    </column>
                                     <property name="name">
                                         <cstring>channelsList</cstring>
                                     </property>
+                                    <property name="sizePolicy">
+                                        <sizepolicy>
+                                            <hsizetype>7</hsizetype>
+                                            <vsizetype>3</vsizetype>
+                                            <horstretch>0</horstretch>
+                                            <verstretch>0</verstretch>
+                                        </sizepolicy>
+                                    </property>
                                     <property name="rootIsDecorated">
                                         <bool>false</bool>
                                     </property>

